### PR TITLE
Fix noise prologue generation

### PR DESCRIPTION
**Pull request by ckousik**

### CODE REVIEW

File: src/sdp.ts [hunk 1]
```diff
@@ -70,12 +93,15 @@ export function fromMultiAddr(ma: Multiaddr, ufrag: string): RTCSessionDescripti
 
 export function munge(desc: RTCSessionDescriptionInit, ufrag: string): RTCSessionDescriptionInit {
   if (desc.sdp) {
-    desc.sdp = desc.sdp
-      .replace(/\na=ice-ufrag:[^\n]*\n/, '\na=ice-ufrag:' + ufrag + '\n')
-      .replace(/\na=ice-pwd:[^\n]*\n/, '\na=ice-pwd:' + ufrag + '\n')
-      ;
-      return desc;
+    desc.sdp = desc.sdp.replace(/\na=ice-ufrag:[^\n]*\n/, '\na=ice-ufrag:' + ufrag + '\n').replace(/\na=ice-pwd:[^\n]*\n/, '\na=ice-pwd:' + ufrag + '\n');
```
Comment (John-LittleBearLabs) [inline src/sdp.ts:96]: This is probably some automated formatter, but I think this is pretty clearly less readable than having the two very similar replace calls aligned.

File: src/sdp.ts [hunk 2]
```diff
@@ -70,12 +93,15 @@ export function fromMultiAddr(ma: Multiaddr, ufrag: string): RTCSessionDescripti
...
+    return desc;
   } else {
     throw new InvalidArgumentError("Can't munge a missing SDP");
   }
 }
+
+export function getCerthashFromMultiaddr(ma: Multiaddr): string | undefined {
```
Comment (John-LittleBearLabs) [inline src/sdp.ts:103]: This appears to be duplicated _some_ of the functionality of `certhash()` ?

File: src/transport.ts [hunk 3]
```diff
@@ -87,15 +91,15 @@ export class WebRTCTransport implements Transport, Initializable {
     let myPeerId = this.components!.getPeerId();
     let rps = ma.getPeerId();
     if (!rps) {
-      throw new Error('TODO Do we really need a peer ID ?');
+      throw new Error('could not get remote peerId');
```
Comment (John-LittleBearLabs) [inline src/transport.ts:94]: This is reworded again in the error PR

File: src/transport.ts [hunk 4]
```diff
@@ -120,4 +124,43 @@ export class WebRTCTransport implements Transport, Initializable {
       remotePeer: theirPeerId,
     });
   }
+
+  private generateNoisePrologue(pc: RTCPeerConnection, ma: Multiaddr): Uint8Array {
+    let remoteCerthash = sdp.getCerthashFromMultiaddr(ma);
+    if (!remoteCerthash) {
+      throw new InvalidArgumentError('no remote tls fingerprint in multiaddr');
```
Comment (John-LittleBearLabs) [inline src/transport.ts:131]: There's an error specific to sending in a multiaddr that's inappropriate for this transport.

**Review by John-LittleBearLabs — APPROVED** [review]

File: src/sdp.ts [hunk 5]
```diff
@@ -39,19 +43,38 @@ function port(ma: Multiaddr): number {
...
-    return certhash_value;
-  } else {
-    throw new Error("Couldn't find a certhash component of multiaddr:" + ma.toString());
+  let certhash_value = tups.filter((tup) => tup[0] == CERTHASH_CODE).map((tup) => tup[1])[0];
+  if (!certhash_value) {
+    throw new InvalidArgumentError('certhash not found in multiaddress');
   }
+
+  // certhash_value is a multibase encoded multihash encoded string
+  // the multiformats PR always encodes in base64
```
[the remark this answers is not in the document]
Comment (mxinden) [reply src/sdp.ts:52]: Correct. As of today, this may be any multibase.

File: src/sdp.ts [hunk 6]
```diff
@@ -39,19 +43,38 @@ function port(ma: Multiaddr): number {
...
+      prefix = 'sha-512';
+      break;
+    default:
+      throw new UnsupportedHashAlgorithmError(mhdecoded.name);
+  }
+
+  let fp = mhdecoded.digest.reduce((str, byte) => str + byte.toString(16).padStart(2, '0'), '');
+  fp = fp.match(/.{1,2}/g)!.join(':');
+
+  return `${prefix} ${fp}`;
```
[the remark this answers is not in the document]
Comment (ckousik) [reply src/sdp.ts:73]: This writes the fingerprint in SDP format `<hash algorithm> <fingerprint>`.

File: src/sdp.ts (hunk 1 above)
Comment (ckousik) [reply src/sdp.ts:96]: This is done by prettier.

File: src/sdp.ts (hunk 2 above)
Comment (ckousik) [reply src/sdp.ts:103]: will fix

File: src/sdp.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ckousik) [reply src/sdp.ts:52]: fixed

**Review by John-LittleBearLabs — APPROVED** [review]

**Pull request merged**